core/src/Kokkos_Graph.hpp

```diff
@@ -0,0 +1,135 @@
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_GRAPH_HPP
+#define KOKKOS_GRAPH_HPP
+
+#include <Kokkos_Macros.hpp>
+#include <impl/Kokkos_Error.hpp>  // KOKKOS_EXPECTS
+
+#include <Kokkos_Graph_fwd.hpp>
+#include <impl/Kokkos_GraphImpl_fwd.hpp>
+
+#include <impl/Kokkos_Utilities.hpp>  // fold emulation
+
+#include <functional>
+#include <memory>
+
+namespace Kokkos {
+namespace Experimental {
+
+//==============================================================================
+// <editor-fold desc="Graph"> {{{1
+
+template <class ExecutionSpace>
+struct KOKKOS_ATTRIBUTE_NODISCARD Graph {
+ public:
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="public member types"> {{{2
+
+  using execution_space = ExecutionSpace;
+  using graph           = Graph;
+  using graph_builder   = GraphBuilder<ExecutionSpace>;
+
+  // </editor-fold> end public member types }}}2
+  //----------------------------------------------------------------------------
+
+ private:
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="friends"> {{{2
+
+  friend struct Kokkos::Impl::GraphAccess;
+
+  // </editor-fold> end friends }}}2
+  //----------------------------------------------------------------------------
+
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="private data members"> {{{2
+
+  using impl_t                       = Kokkos::Impl::GraphImpl<ExecutionSpace>;
+  std::shared_ptr<impl_t> m_impl_ptr = nullptr;
+
+  // </editor-fold> end private data members }}}2
+  //----------------------------------------------------------------------------
+
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="private ctors"> {{{2
+
+  // Note: only create_graph() uses this constructor, but we can't just make
+  // that a friend instead of GraphAccess because of the way that friend
+  // function template injection works.
+  explicit Graph(std::shared_ptr<impl_t> arg_impl_ptr)
+      : m_impl_ptr(std::move(arg_impl_ptr)) {}
+
+  // </editor-fold> end private ctors }}}2
+  //----------------------------------------------------------------------------
+
+ public:
+  ExecutionSpace const& get_execution_space() const {
+    return m_impl_ptr->get_execution_space();
+  }
+
+  void submit() const& {
+    KOKKOS_EXPECTS(bool(m_impl_ptr))
+    (*m_impl_ptr).submit();
+  }
+
+  void submit() && {
+    KOKKOS_EXPECTS(bool(m_impl_ptr))
+    // The graph interface isn't thread-safe, so we can rely on this
+    if (m_impl_ptr.use_count() == 1) {
```

Comment:
Nonblocking:  Using `use_count() == 1` assumes that no other threads have weak_ptrs to it.  Do we know that to be true?

Reply:
Yeah, that's what I was trying to say with this comment: 

```
// The graph interface isn't thread-safe, so we can rely on this
```